src/storage.zig

```diff
@@ -7,400 +7,555 @@ const log = std.log.scoped(.storage);
+                // careful to avoid logging in a busy loop. Perhaps a better approach might be to
+                // return wrote = null here and let the protocol retry at a higher layer where
+                // there is more context available to decide on how important this is or whether
+                // to cancel.
+                @panic("write operation returned 0 bytes written");
+            }
+
+            write.offset += bytes_written;
+            write.buffer = write.buffer[bytes_written..];
+
+            if (write.buffer.len == 0) {
+                write.callback(write);
+                return;
+            }
+
+            self.start_write(write);
         }
-    }
 
-    pub fn read_sectors(
-        self: *Storage,
-        callback: *const fn (read: *Storage.Read) void,
-        read: *Storage.Read,
-        buffer: []u8,
-        zone: vsr.Zone,
-        offset_in_zone: u64,
-    ) void {
-        zone.verify_iop(buffer, offset_in_zone);
-        assert(zone != .grid_padding);
-
-        const offset_in_storage = zone.offset(offset_in_zone);
-        read.* = .{
-            .completion = undefined,
-            .callback = callback,
-            .buffer = buffer,
-            .offset = offset_in_storage,
-            .target_max = buffer.len,
-        };
+        /// Ensures that the read or write is within bounds and intends to read or write some bytes.
+        fn assert_bounds(self: *Self, buffer: []const u8, offset: u64) void {
+            _ = self;
+            _ = offset;
 
-        self.start_read(read, null);
-        assert(read.target().len > 0);
-    }
+            assert(buffer.len > 0);
+        }
+    };
+}
 
-    fn start_read(self: *Storage, read: *Storage.Read, bytes_read: ?usize) void {
-        assert(read.offset % constants.sector_size == 0);
-        maybe(bytes_read == 0); // Retrying erroneous read at the same offset with smaller window.
+test "storage with mocked io layer" {
+    const IO = @import("testing/io.zig").IO;
+    const StorageWithMockedIO = Storage(IO);
 
-        const bytes = bytes_read orelse 0;
-        assert(bytes <= read.target().len);
+    const seed: u64 = 0;
 
-        read.offset += bytes;
-        read.buffer = read.buffer[bytes..];
+    const zones: []const vsr.Zone = &.{
+        .superblock,
+        .wal_headers,
+        .wal_prepares,
+        .client_replies,
+    };
 
-        const target = read.target();
-        if (target.len == 0) {
-            // Resolving the read inline means start_read() must not have been called from
-            // read_sectors(). If it was, this is a synchronous callback resolution and should
-            // be reported.
-            assert(bytes_read != null);
+    const storage_size = constants.sector_size * 64;
 
-            read.callback(read);
-            return;
-        }
+    var written_storage_data: [storage_size]u8 align(constants.sector_size) = undefined;
```

Comment:
We prefer big endian naming (`storage_data_written`, `storage_data_stored` etc) so that the prefixes can line up!